File: ktor-frotagestor/src/main/kotlin/interfaces/Vehicle.kt
package com.frotagestor.interfaces

import kotlinx.serialization.Serializable

@Serializable
data class Vehicle(
    val id: Int? = null,
    val plate: String,
    val model: String,
    val brand: String? = null,
    val year: Int? = null,
    val status: VehicleStatus = VehicleStatus.ATIVO
)

@Serializable
enum class VehicleStatus {
    ATIVO,
    INATIVO,
    MANUTENCAO
}

@Serializable
data class PartialVehicle(
    val id: Int? = null,
    val plate: String? = null,
    val model: String? = null,
    val brand: String? = null,
    val year: Int? = null,
    val status: VehicleStatus = VehicleStatus.ATIVO
)

@Serializable
data class VehicleIndicators(
    val active: Int,
    val maintenance: Int,
    val lastVehicle: Vehicle? = null
)

@Serializable
data class VehicleReport(
    val distributions: Distributions,
    val usageStats: UsageStats
) {
    @Serializable
    data class Distributions(
        val byBrand: List<ByBrand>,
        val byYear: List<ByYear>,
        val byStatus: List<ByStatus>
    ) {
        @Serializable
        data class ByBrand(
            val brand: String,
            val count: Long
        )

        @Serializable
        data class ByYear(
            val year: Int,
            val count: Long
        )

        @Serializable
        data class ByStatus(
            val status: Status,
            val count: Long
        )

        @Serializable
        enum class Status {
            ATIVO,
            MANUTENCAO,
            INATIVO
        }
    }

    @Serializable
    data class UsageStats(
        val totalDistanceByVehicle: List<TotalDistanceByVehicle>,
        val fuelConsumptionByVehicle: List<FuelConsumptionByVehicle>
    ) {
        @Serializable
        data class TotalDistanceByVehicle(
            val plate: String,
            val totalKm: Int,
            val totalTrips: Long,
            val topDriver: TopDriver? = null,
            val fuelCost: Double,
            val maintenanceCost: Double,
            val totalCost: Double,
            val lastMaintenanceDate: String? = null,
            val isInUse: Boolean? = null
        ) {
            @Serializable
            data class TopDriver(
                val name: String?,
                val trips: Int
            )
        }

        @Serializable
        data class FuelConsumptionByVehicle(
            val plate: String,
            val litersPerKm: Double
        )
    }
}

File: frontend/src/app/interfaces/vehicle.ts
// vehicle.ts
export interface Vehicle {
  id?: number;
  plate: string;
  model: string;
  brand?: string | null;
  year?: number | null;
  status: VehicleStatus;
  iconMapUrl: string;
}

export enum VehicleStatus {
  ATIVO = 'ATIVO',
  INATIVO = 'INATIVO',
  MANUTENCAO = 'MANUTENCAO',
}

export interface VehicleIndicators {
  active: number;
  maintenance: number;
  lastVehicle: {
    plate: string;
    date: string;
  };
}

export interface VehicleReport {
  distributions: {
    byBrand: {
      brand: string;
      count: number;
    }[];
    byYear: {
      year: number;
      count: number;
    }[];
    byStatus: {
      status: 'ATIVO' | 'MANUTENCAO' | 'INATIVO';
      count: number;
    }[];
  };

  usageStats: {
    totalDistanceByVehicle: {
      plate: string;
      totalKm: number;
      totalTrips: number; // total de viagens lançadas para esse veículo
      topDriver?: {
        name: string;
        trips: number;
      }; // motorista que mais usou o veículo
      fuelCost: number; // gasto total com combustível
      maintenanceCost: number; // gasto total com manutenção
      totalCost: number; // gasto total do veículo
      lastMaintenanceDate?: string; // data da manutenção mais recente
      isInUse?: boolean; // true se o veículo estiver em movimento no momento
    }[];
    fuelConsumptionByVehicle: {
      plate: string;
      litersPerKm: number;
    }[];
  };
}
